lib/bundler/fetcher.rb

```diff
@@ -374,35 +371,35 @@ def bundler_cert_store
       store
     end
 
-    # Attempt to retry with HTTP authentication, if it's appropriate to do so. Yields to a block;
-    # the caller should use this to re-attempt the failing request with the altered `@remote_uri`.
-    def retry_with_auth
-      # Authentication has already been attempted and failed.
-      raise BadAuthenticationError.new(uri) if @remote_uri.user
-
-      auth = Bundler.settings[@remote_uri.to_s]
-
-      # Authentication isn't provided at all, by "bundle config" or in the URI.
-      raise AuthenticationRequiredError.new(uri) if auth.nil?
+  private
 
-      @remote_uri.user, @remote_uri.password = *auth.split(":", 2)
-      @anonymizable_uri = AnonymizableURI.new(@remote_uri.dup)
-      yield
+    def resolve_remote_uri(uri)
+      add_configured_credentials(Bundler::Source.mirror_for(uri))
     end
 
-  private
+    def add_configured_credentials(uri)
+      auth = Bundler.settings[uri.to_s]
+      if auth
+        uri = uri.dup
+        uri.user, uri.password = *auth.split(":", 2)
```

Comment:
Hmm... I just realised that this will now make credentials in config override ones in the `Gemfile`.

We never really clearly defined what would happen when you have both, but in the past I suppose it would have tried the ones in the `Gemfile`, then fail immediately if the original ones didn't work (it got a 401 or 403 response).

I guess, intuitively, I would expect that the ones in the `Gemfile` would override the ones in config, but what do other people think?